compiler/src/dotty/tools/dotc/transform/TreeMapWithStages.scala

```diff
@@ -0,0 +1,145 @@
+package dotty.tools.dotc
+package transform
+
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.ast.{TreeMapWithImplicits, TreeTypeMap, tpd, untpd}
+import dotty.tools.dotc.core.Constants._
+import dotty.tools.dotc.core.Decorators._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.quoted._
+import dotty.tools.dotc.core.NameKinds._
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.StdNames._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.tasty.TreePickler.Hole
+import dotty.tools.dotc.util.Spans._
+import dotty.tools.dotc.util.{Property, SourcePosition}
+import dotty.tools.dotc.transform.SymUtils._
+import dotty.tools.dotc.typer.Implicits.SearchFailureType
+import dotty.tools.dotc.typer.Inliner
+
+import scala.collection.mutable
+import scala.annotation.constructorOnly
+
+/** The main transformer class
+ *  @param  level      the current level, where quotes add one and splices subtract one level.
+ *                     The initial level is 0, a level `l` where `l > 0` implies code has been quoted `l` times
+ *                     and `l == -1` is code inside a top level splice (in an inline method).
+ *  @param  levels     a stacked map from symbols to the levels in which they were defined
+ */
+abstract class TreeMapWithStages(@constructorOnly ictx: Context) extends TreeMapWithImplicits {
+  import tpd._
+  import TreeMapWithStages._
+
+  /** A map from locally defined symbols to their definition quotation level */
+  private[this] val levelOfMap: mutable.HashMap[Symbol, Int] = ictx.property(LevelOfKey).get
+
+  /** A stack of entered symbols, to be unwound after scope exit */
+  private[this] var enteredSyms: List[Symbol] = Nil
+
+  /** The quotation level of the definition of the locally defined symbol */
+  protected def levelOf(sym: Symbol): Option[Int] = levelOfMap.get(sym)
+
+  /** Localy defined symbols seen so far by `StagingTransformer.transform` */
+  protected def localSymbols: List[Symbol] = enteredSyms
+
+  /** Enter staging level of symbol defined by `tree`, if applicable. */
+  private def markDef(tree: Tree)(implicit ctx: Context): Unit = tree match {
+    case tree: DefTree =>
+      val sym = tree.symbol
+      if ((sym.isClass || !sym.maybeOwner.isType) && !levelOfMap.contains(sym)) {
```

Comment:
`sym.maybeOwber.isTerm` instead of `!sym.maybeOwner.isType`?

Reply:
Yes